pkg/function/backup_data_using_kopia_server.go

```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch Kopia API Server Certificate Secret Data from blueprint")
+	}
+
+	username := tp.RepositoryServer.Username
+	hostname, userAccessPassphrase, err := getHostNameAndUserPassPhraseFromRepoServer(userPassphrase)
+
+	cli, err := kube.NewClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create Kubernetes client")
+	}
+
+	ctx = field.Context(ctx, consts.PodNameKey, pod)
+	ctx = field.Context(ctx, consts.ContainerNameKey, container)
+
+	repositoryServerService, err := cli.CoreV1().Services(tp.RepositoryServer.Namespace).Get(context.Background(), tp.RepositoryServer.ServerInfo.ServiceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.New("Unable to find Service Details for Repository Server")
+	}
+	repositoryServerServicePort := strconv.Itoa(int(repositoryServerService.Spec.Ports[0].Port))
+	serverAddress := "https://" + tp.RepositoryServer.ServerInfo.ServiceName + "." + tp.RepositoryServer.Namespace + ".svc.cluster.local:" + repositoryServerServicePort
+
+	snapInfo, err := backupDataUsingKopiaServer(
+		cli,
+		container,
+		hostname,
+		includePath,
+		namespace,
+		pod,
+		serverAddress,
+		fingerprint,
+		username,
+		userAccessPassphrase,
+		tags,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to backup data to Kopia API server")
+	}
+
+	var logSize, phySize int64
+	if snapInfo.Stats != nil {
+		stats := snapInfo.Stats
+		logSize = stats.SizeHashedB + stats.SizeCachedB
+		phySize = stats.SizeUploadedB
+	}
+
+	output := map[string]any{
+		BackupDataOutputBackupID:           snapInfo.SnapshotID,
+		BackupDataOutputBackupSize:         humanize.Bytes(uint64(logSize)),
+		BackupDataOutputBackupPhysicalSize: humanize.Bytes(uint64(phySize)),
+	}
+	return output, nil
+}
+
+func backupDataUsingKopiaServer(
+	cli kubernetes.Interface,
+	container,
+	hostname,
+	includePath,
+	namespace,
+	pod,
+	serverAddress,
+	fingerprint,
+	username,
+	userPassphrase string,
+	tags []string,
+) (info *kopiacmd.SnapshotCreateInfo, err error) {
+	contentCacheMB, metadataCacheMB := kopiacmd.GetCacheSizeSettingsForSnapshot()
+	configFile, logDirectory := kankopia.GetCustomConfigFileAndLogDirectory(hostname)
+
+	cmd := kopiacmd.RepositoryConnectServerCommand(kopiacmd.RepositoryServerCommandArgs{
+		UserPassword:    userPassphrase,
+		ConfigFilePath:  configFile,
+		LogDirectory:    logDirectory,
+		CacheDirectory:  kopiacmd.DefaultCacheDirectory,
+		Hostname:        hostname,
+		ServerURL:       serverAddress,
+		Fingerprint:     fingerprint,
+		Username:        username,
+		ContentCacheMB:  contentCacheMB,
+		MetadataCacheMB: metadataCacheMB,
+	})
+
+	stdout, stderr, err := kube.Exec(cli, namespace, pod, container, cmd, nil)
+	format.Log(pod, container, stdout)
+	format.Log(pod, container, stderr)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to connect to Kopia API server")
+	}
+
+	cmd = kopiacmd.SnapshotCreate(
```

Comment:
how is this command using the serverAddress that eventually is the repo server's service?

Reply:
This command specifically doesn't use Repository Server Address, if you look above on line 180, we have ```RepositoryConnectServerCommand```, first we make the connection to Kopia Repo Server, that requires the Repo Server Address, than we run snapshot command which than doesn't requires it

Reply:
ok, so I am assuming somehow after connection is created the same connection is used by the snapshot create command/snapshot?

Reply:
Yes @viveksinghggits

Reply:
how?

Reply:
According to my understanding, the connection details are being taken from the configFile. @ankitjain235 correct me if I'm wrong.